fix: Raise 500 exception instead of returning an error on `detect` API call

ramyma opened the pull request:
I noticed that the `detect` API call returns with a `200` status code when the selected module is not implemented/available or no when no input image is provided.
This PR raises an 500 status code HTTP exception for better integration with 3rd party implementations.

Note: I couldn't find a PEP formatting guide on the repo, so I didn't format the file on saving to avoid having unrelated changes on the PR.

huchenlei commented:
Can you add an unittest under `tests/web_api` for this change? Thanks!

ramyma commented:
Added tests and rebased on main branch.

ljleb commented on scripts/api.py:
```diff
@@ -205,9 +206,59 @@ async def detect(
+            raise HTTPException(
+                status_code=500, detail="Module not available")
+        
+        available_modules = [
```
is there a reason we need to duplicate the preprocessors keys here? The exception above should handle all cases.

ljleb replied:
Ah it seems to be an artifact from older code. In any case, please consider removing it.

ramyma replied:
You're right, it was pulled in with the rebase for some reason!
I'll remove it.

ramyma commented:
I rebased again with the latest and fixed a couple of mistakes in the tests.

I wish there were an easier way to run the tests as it takes quite a long time to spin up the whole webui environment through `launch.py`.

ljleb commented on tests/web_api/detect_test.py:
```diff
@@ -0,0 +1,39 @@
+        self.assertEqual(utils.detect(detect_args).status_code, 500)
+
+    def test_detect_with_no_input_images_performed(self):
+        detect_args = self.base_detect_args.update({
```
update returns `None` and updates the dict in place. Maybe just use `self.base_detect_args` directly in the assert below instead.

ramyma replied:
Looks like my python is a bit rusty :)
I've switched it up to handle it without mutating the base args.

ljleb commented on tests/web_api/detect_test.py:
```diff
@@ -0,0 +1,39 @@
+        }
+
+    def test_detect_with_invalid_module_performed(self):
+        detect_args = self.base_detect_args | {
```
Although this is a good solution for python 3.10, this makes the test not work under python 3.8 as the dict union operator was introduced in python 3.9. It was discussed earlier that the repo should support both 3.8 and 3.10.

As this is code that will only be run during automated testing, this may be okay. However, this means we now can only run the tests using 3.10 now.

ramyma replied:
I wasn't aware the repo should support 3.8 as well.
I pushed another update to cover it.
It would be nice if there's a contribution guide, maybe it's there already and I missed.

ljleb replied:
No problem, thanks for your time updating the code.

ljleb commented on tests/web_api/detect_test.py:
```diff
@@ -0,0 +1,39 @@
+        }
+
+    def test_detect_with_invalid_module_performed(self):
+        detect_args = self.base_detect_args.copy().update({
```
```suggestion
        detect_args = self.base_detect_args.copy()
        detect_args.update({
```

ljleb commented on tests/web_api/detect_test.py:
```diff
@@ -0,0 +1,39 @@
+        self.assertEqual(utils.detect(detect_args).status_code, 500)
+
+    def test_detect_with_no_input_images_performed(self):
+        detect_args = self.base_detect_args.copy().update({
```
```suggestion
        detect_args = self.base_detect_args.copy()
        detect_args.update({
```

ramyma commented:
@ljleb  I was just about to push the same changes that you pushed.
It's kinda late and I missed it :)
Thanks for the support!

ljleb commented:
Ah apologies, I didn't see the changes were recent. Should have waited.

Would it be possible to use http code 422 instead of 500? 500 is usually more employed for internal server errors, as opposed to bad request parameters or body.

ljleb commented:
It may be a good time to add a contributing.md file so that contributors know what to do and we don't have to tell everybody these things all the time...

ramyma commented:
Done!

ljleb commented:
Thanks for your contribution! I'll test the code now.
